File: orphaned-categories-outside-root-tree/node/find-orphaned-categories.js
/**
 * Find PrestaShop categories and products orphaned outside the root tree.
 *
 * PrestaShop stores categories as a nested set tree (id_parent plus internal
 * nleft/nright bounds) rooted at each shop's designated root category
 * (shops.id_category, typically Home under a hidden super-root). If the root is
 * deleted directly instead of through the shop's reassignment flow, or a
 * category or product import sets id_parent to a non-existent or wrong-shop id,
 * child categories keep an id_parent that no longer resolves back to the root.
 * The front office only renders nodes reachable from the root, so the row stays
 * active in ps_category, and products stay linked via ps_category_product, but
 * neither is visible anywhere.
 *
 * This script reads each shop's true root id, pulls every category and active
 * product over the webservice, walks id_parent links with a breadth first
 * search from the root, and runs a pure decision function that flags any
 * category or product the walk never reaches. It reports by default. A
 * corrective PUT that re-parents an orphaned category root to the shop's Home
 * category is only sent when DRY_RUN=false and the target has been confirmed.
 *
 * Guide: https://www.allanninal.dev/prestashop/orphaned-categories-outside-root-tree/
 *
 * Run on a schedule, or right after a suspicious import. Safe to run again and again.
 */
import { pathToFileURL } from "node:url";

const PRESTASHOP_URL = (process.env.PRESTASHOP_URL || "https://demo.example.com").replace(/\/+$/, "");
const PRESTASHOP_WS_KEY = process.env.PRESTASHOP_WS_KEY || "WSKEYDUMMY";
const DRY_RUN = (process.env.DRY_RUN || "true").toLowerCase() === "true";

function basicAuthHeader() {
  return "Basic " + Buffer.from(`${PRESTASHOP_WS_KEY}:`).toString("base64");
}

/**
 * Pure decision function, no I/O.
 *
 * categories: array of plain objects, each with id, id_parent, is_root_category.
 * rootIds: Set<number> or array of valid shop root category ids (shops.id_category).
 * products: array of plain objects, each with id, id_category_default, category_ids (array<number>).
 *
 * Builds a parent to children adjacency map, walks it with a breadth first
 * search from rootIds to compute the reachable category ids, then returns the
 * category ids and product ids that walk never reaches.
 *
 * Returns { orphaned_categories: number[], orphaned_products: number[] }.
 */
export function findOrphans(categories, rootIds, products) {
  const rootSet = rootIds instanceof Set ? rootIds : new Set(rootIds);
  const children = new Map();
  for (const cat of categories) {
    const parent = cat.id_parent;
    if (parent !== null && parent !== undefined) {
      if (!children.has(parent)) children.set(parent, []);
      children.get(parent).push(cat.id);
    }
  }

  const reachable = new Set(rootSet);
  const queue = [...rootSet];
  while (queue.length) {
    const current = queue.shift();
    for (const childId of children.get(current) || []) {
      if (!reachable.has(childId)) {
        reachable.add(childId);
        queue.push(childId);
      }
    }
  }

  const orphanedCategories = categories
    .filter((cat) => !reachable.has(cat.id) && !rootSet.has(cat.id))
    .map((cat) => cat.id);

  const orphanedProducts = products
    .filter((p) => {
      const defaultReachable = reachable.has(p.id_category_default);
      const anyReachable = (p.category_ids || []).some((cid) => reachable.has(cid));
      return !defaultReachable && !anyReachable;
    })
    .map((p) => p.id);

  return { orphaned_categories: orphanedCategories, orphaned_products: orphanedProducts };
}

async function apiGet(path, params = {}) {
  const url = new URL(`${PRESTASHOP_URL}/api/${path}`);
  url.searchParams.set("output_format", "JSON");
  for (const [key, value] of Object.entries(params)) url.searchParams.set(key, value);
  const res = await fetch(url, { headers: { Authorization: basicAuthHeader() } });
  if (!res.ok) throw new Error(`PrestaShop ${res.status} on GET ${path}`);
  return res.json();
}

async function apiPut(path, resourceKey, body, params = {}) {
  const url = new URL(`${PRESTASHOP_URL}/api/${path}`);
  url.searchParams.set("output_format", "JSON");
  for (const [key, value] of Object.entries(params)) url.searchParams.set(key, value);
  const res = await fetch(url, {
    method: "PUT",
    headers: { Authorization: basicAuthHeader(), "Content-Type": "application/json" },
    body: JSON.stringify({ [resourceKey]: body }),
  });
  if (!res.ok) throw new Error(`PrestaShop ${res.status} on PUT ${path}`);
  return res.json();
}

async function shopRootIds() {
  const data = await apiGet("shops", { display: "full" });
  const rows = data.shops || [];
  const roots = new Set(rows.filter((r) => r.id_category).map((r) => Number(r.id_category)));
  if (roots.size) return roots;
  const cfg = await apiGet("configurations", { "filter[name]": "PS_HOME_CATEGORY" });
  const cfgRows = cfg.configurations || [];
  return new Set(cfgRows.filter((r) => r.value).map((r) => Number(r.value)));
}

async function allCategories() {
  const data = await apiGet("categories", { display: "full", limit: "0" });
  const rows = data.categories || [];
  return rows.map((row) => ({
    id: Number(row.id),
    id_parent: row.id_parent !== undefined && row.id_parent !== null && row.id_parent !== "" ? Number(row.id_parent) : null,
    is_root_category: ["1", "true", true].includes(row.is_root_category),
  }));
}

async function allActiveProducts() {
  const data = await apiGet("products", { display: "full", "filter[active]": "1", limit: "0" });
  const rows = data.products || [];
  return rows.map((row) => {
    const cats = (row.associations && row.associations.categories && row.associations.categories.category) || [];
    const categoryIds = cats.filter((c) => c.id).map((c) => Number(c.id));
    const hasDefault = row.id_category_default !== undefined && row.id_category_default !== null && row.id_category_default !== "";
    return {
      id: Number(row.id),
      id_category_default: hasDefault ? Number(row.id_category_default) : null,
      category_ids: categoryIds,
    };
  });
}

async function reparentCategoryToHome(category, homeCategoryId) {
  // Only used when DRY_RUN=false and a safe target has been confirmed.
  // PrestaShop recomputes nleft/nright for the moved subtree on save.
  const body = { ...category, id_parent: homeCategoryId };
  return apiPut(`categories/${category.id}`, "category", body);
}

export async function run() {
  const rootIds = await shopRootIds();
  const categories = await allCategories();
  const products = await allActiveProducts();
  const result = findOrphans(categories, rootIds, products);

  const { orphaned_categories: orphanedCategories, orphaned_products: orphanedProducts } = result;
  const byId = new Map(categories.map((cat) => [cat.id, cat]));

  for (const catId of orphanedCategories) {
    const cat = byId.get(catId) || {};
    console.warn(`Orphaned category id=${catId} id_parent=${cat.id_parent}`);
  }

  for (const prodId of orphanedProducts) {
    console.warn(`Orphaned product id=${prodId}`);
  }

  if (!DRY_RUN && orphanedCategories.length && rootIds.size) {
    const homeId = [...rootIds][0];
    for (const catId of orphanedCategories) {
      await reparentCategoryToHome(byId.get(catId), homeId);
      console.log(`Re-parented category id=${catId} to Home id_parent=${homeId}.`);
    }
  }

  console.log(
    `Done. ${orphanedCategories.length} orphaned categorie(s), ${orphanedProducts.length} orphaned product(s) ${DRY_RUN ? "reported" : "reported and categories repaired"}.`
  );
}

if (import.meta.url === pathToFileURL(process.argv[1]).href) {
  run().catch((err) => { console.error(err); process.exit(1); });
}

File: orphaned-categories-outside-root-tree/python/find_orphaned_categories.py
"""Find PrestaShop categories and products orphaned outside the root tree.

PrestaShop stores categories as a nested set tree (id_parent plus internal
nleft/nright bounds) rooted at each shop's designated root category
(shops.id_category, typically Home under a hidden super-root). If the root is
deleted directly instead of through the shop's reassignment flow, or a
category or product import sets id_parent to a non-existent or wrong-shop id,
child categories keep an id_parent that no longer resolves back to the root.
The front office only renders nodes reachable from the root, so the row stays
active in ps_category, and products stay linked via ps_category_product, but
neither is visible anywhere.

This script reads each shop's true root id, pulls every category and active
product over the webservice, walks id_parent links with a breadth first
search from the root, and runs a pure decision function that flags any
category or product the walk never reaches. It reports by default. A
corrective PUT that re-parents an orphaned category root to the shop's Home
category is only sent when DRY_RUN=false and the target has been confirmed.

Guide: https://www.allanninal.dev/prestashop/orphaned-categories-outside-root-tree/

Run on a schedule, or right after a suspicious import. Safe to run again and again.
"""
import os
import logging
from collections import deque
import requests

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("find_orphaned_categories")

PRESTASHOP_URL = os.environ.get("PRESTASHOP_URL", "https://demo.example.com").rstrip("/")
PRESTASHOP_WS_KEY = os.environ.get("PRESTASHOP_WS_KEY", "WSKEYDUMMY")
DRY_RUN = os.environ.get("DRY_RUN", "true").lower() == "true"
AUTH = (PRESTASHOP_WS_KEY, "")


def find_orphans(categories, root_ids, products):
    """Pure decision function, no I/O.

    categories: list[dict] each with id, id_parent, is_root_category.
    root_ids: set[int] of valid shop root category ids (shops.id_category).
    products: list[dict] each with id, id_category_default, category_ids (list[int]).

    Builds a parent to children adjacency map, walks it with a breadth first
    search from root_ids to compute reachable_category_ids, then returns the
    category ids and product ids that walk never reaches.

    Returns {"orphaned_categories": [...], "orphaned_products": [...]}.
    """
    root_ids = set(root_ids)

    children = {}
    for cat in categories:
        parent = cat.get("id_parent")
        if parent is not None:
            children.setdefault(parent, []).append(cat["id"])

    reachable = set(root_ids)
    queue = deque(root_ids)
    while queue:
        current = queue.popleft()
        for child_id in children.get(current, []):
            if child_id not in reachable:
                reachable.add(child_id)
                queue.append(child_id)

    orphaned_categories = [
        cat["id"] for cat in categories
        if cat["id"] not in reachable and cat["id"] not in root_ids
    ]

    orphaned_products = [
        p["id"] for p in products
        if p.get("id_category_default") not in reachable
        and not any(cid in reachable for cid in p.get("category_ids") or [])
    ]

    return {"orphaned_categories": orphaned_categories, "orphaned_products": orphaned_products}


def api_get(path, params=None):
    params = dict(params or {})
    params["output_format"] = "JSON"
    r = requests.get(f"{PRESTASHOP_URL}/api/{path}", params=params, auth=AUTH, timeout=30)
    r.raise_for_status()
    return r.json()


def api_put(path, resource_key, body, params=None):
    params = dict(params or {})
    params["output_format"] = "JSON"
    r = requests.put(
        f"{PRESTASHOP_URL}/api/{path}",
        params=params, auth=AUTH,
        json={resource_key: body}, timeout=30,
    )
    r.raise_for_status()
    return r.json()


def shop_root_ids():
    data = api_get("shops", params={"display": "full"})
    rows = data.get("shops") or []
    roots = {int(row["id_category"]) for row in rows if row.get("id_category")}
    if roots:
        return roots
    cfg = api_get("configurations", params={"filter[name]": "PS_HOME_CATEGORY"})
    rows = cfg.get("configurations") or []
    return {int(row["value"]) for row in rows if row.get("value")}


def all_categories():
    data = api_get("categories", params={"display": "full", "limit": "0"})
    rows = data.get("categories") or []
    return [
        {
            "id": int(row["id"]),
            "id_parent": int(row["id_parent"]) if row.get("id_parent") not in (None, "") else None,
            "is_root_category": str(row.get("is_root_category")) in ("1", "true", "True"),
        }
        for row in rows
    ]


def all_active_products():
    data = api_get("products", params={"display": "full", "filter[active]": "1", "limit": "0"})
    rows = data.get("products") or []
    products = []
    for row in rows:
        cats = ((row.get("associations") or {}).get("categories") or {}).get("category") or []
        category_ids = [int(c["id"]) for c in cats if c.get("id")]
        default = row.get("id_category_default")
        products.append({
            "id": int(row["id"]),
            "id_category_default": int(default) if default not in (None, "") else None,
            "category_ids": category_ids,
        })
    return products


def reparent_category_to_home(category, home_category_id):
    # Only used when DRY_RUN=false and a safe target has been confirmed.
    # PrestaShop recomputes nleft/nright for the moved subtree on save.
    body = dict(category)
    body["id_parent"] = home_category_id
    return api_put(f"categories/{category['id']}", "category", body)


def run():
    root_ids = shop_root_ids()
    categories = all_categories()
    products = all_active_products()
    result = find_orphans(categories, root_ids, products)

    orphaned_categories = result["orphaned_categories"]
    orphaned_products = result["orphaned_products"]
    by_id = {cat["id"]: cat for cat in categories}

    for cat_id in orphaned_categories:
        cat = by_id.get(cat_id, {})
        log.warning("Orphaned category id=%s id_parent=%s", cat_id, cat.get("id_parent"))

    for prod_id in orphaned_products:
        log.warning("Orphaned product id=%s", prod_id)

    if not DRY_RUN and orphaned_categories and root_ids:
        home_id = next(iter(root_ids))
        for cat_id in orphaned_categories:
            reparent_category_to_home(by_id[cat_id], home_id)
            log.info("Re-parented category id=%s to Home id_parent=%s.", cat_id, home_id)

    log.info(
        "Done. %d orphaned categorie(s), %d orphaned product(s) %s.",
        len(orphaned_categories), len(orphaned_products),
        "reported" if DRY_RUN else "reported and categories repaired",
    )


if __name__ == "__main__":
    run()
